src/compiler/checker.ts

```diff
@@ -32423,22 +32423,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getContextualCallSignature(type, node);
         }
         let signatureList: Signature[] | undefined;
-        const types = (type as UnionType).types;
+        const types = (type as UnionType).types
+            .map(type => ({ type, signature: getContextualCallSignature(type, node) }))
+            .filter(type => type.signature)
+            .sort((t1, t2) => t1.signature!.parameters.length - t2.signature!.parameters.length);

```

Comment:
You need to account for rest parameters here so it would be more appropriate to use `getParameterCount` here over directly checking the `.parameters.length`

Reply:
Thank you. I think that actually what I need is `getMinArgumentCount`.
For the case `((a: string, b?: string) => void) | ((a: string, ...rest: string[]) => void)` I don't think it matters which one is first.

Reply:
I'm not sure if it will make any difference but to exercise even more funky test cases you could add test cases with signatures of this shape:
```ts
((a: string, ...rest: [string, number?] | [number, boolean?]) => void) | ((a: string, ...rest: [string, number] | [number, boolean]) => void)
```

Note that I didn't think through what types you should use in those rest parameters, just that they can be unions of tuples and that one of the signatures might have some of the elements there required and some might have them optional etc.